### PR TITLE
IMPROVE HMR error reporting, no refreshes needed for error recovery

### PR DESCRIPTION
**Pull request by ndelangen**
Issue: https://github.com/storybookjs/storybook/issues/5822

## What I did
I fixed the HMR for error handling and improved the display of errors a bit

## How to test
- open any example
- do a code-change in a story file
- assert the codechange is applied without a refresh
- do a code-change with a syntax error
- assert there is a error message
- navigate to another story
- assert that the error message stays in view
- fix the syntax error
- assert that the error disappears and you will navigate to the correct story

also:
- throw an error in the storyFn
- assert that the error is displayed
- fix the error in the story
- assert that the story renders correctly without a refresh

Here's how the error look now:
Screenshot 2019-06-04 at 21 55 09
Screenshot 2019-06-04 at 21 54 58

### CODE REVIEW

**Comment by ndelangen** [comment]
By it's nature of HMR only working and doing anything during development, this PR should be tested manually on multiple machines!

**Comment by tmeasday** [comment]
Wow, this is cool @ndelangen. I'd like to take some time to test it out properly

File: lib/core/src/client/preview/start.js [hunk 1]
```diff
@@ -118,13 +122,17 @@ export default function start(render, { decorateStory } = {}) {
...
 
   const renderMain = forceRender => {
     const revision = storyStore.getRevision();
     const selection = storyStore.getSelection();
-    const { kind, name, getDecorated, id } = selection || {};
+    const loadError = storyStore.getError();
+    const { kind, name, getDecorated, id, error } = selection || {};
 
-    if (getDecorated) {
+    if (loadError || error || (revision === previousRevision && previousError)) {
```
Comment (tmeasday) [inline lib/core/src/client/preview/start.js]: What is the situation where `previousError` is defined but `error` is not?

Comment (tmeasday) [reply lib/core/src/client/preview/start.js]: I don't even see `previousError` getting set anywhere...?

**Review by tmeasday — COMMENTED** [review]
Code seems general good (one comment) but I'd like to try this out a bit.

File: lib/core/src/client/preview/start.js (hunk 1 above)
Comment (ndelangen) [reply lib/core/src/client/preview/start.js]: good catch, will fix this

**Comment by ndelangen** [comment]
@tmeasday were you able to give this a test locally?

**Comment by tmeasday** [comment]
Oh, sorry I will try and do it today

**Comment by tmeasday** [comment]
Works great @ndelangen. PS I noticed that our `errors` story in official storybook (http://localhost:9011/?path=/story/core-errors--story-errors) -- where we are supposed to return something that React cannot render no longer works. I think this is because React can now render `null`?

Should we change it to `export const errors = () => ({});` -- react doesn't like that.

**Comment by ndelangen** [comment]
Exciting! possibly a headline feature @shilman ?

I'll fix up the merge conflicts & make sure that story you mentioned actually throws an error.

**Comment by ndelangen** [comment]
@tmeasday if a storyFn returns something that react can't render, it doesn't seem to do anything, it just renders nothing in the place. I can't find a way to make that actually throw anymore. React just seems to accept it 🤷‍♂ 

This is because the storyFn isn't executed in the react render stack, and so no InvariantError occurs.

When a component rendered in a story tries to do the same though, it DOES throw a InvariantError.

Would be great to address this at some point, but that doesn't feel like a stopper for this PR.

**Comment by tmeasday** [comment]
Sure thing @ndelangen

**Pull request merged**